docker-nginx/templates/default.conf.template

```diff
@@ -25,13 +25,19 @@ log_format json-logger escape=json
   '"source":"nginx"'
 '}';
 
+upstream django {
+  # Defines a shared memory zone between worker processes, necessary dynamic dns resolving
+  zone django 64k;
 
+  # Add Docker's DNS resolver with a short TTL and IPv6 turned off
+  resolver 127.0.0.11 valid=1s ipv6=off;
 
-upstream django {
   # the number of keepalive connections comes from the number Django workers we have
   keepalive 32;
 
-  server app:8000 fail_timeout=0;
+  # Use Docker DNS resolution to dynamically resolve the app backend hostname
+  # max_fails=1 and fail_timeout=1s ensures Nginx retries on almost every request from loading.html
```

Comment:
```suggestion
  # `max_fails=1` and `fail_timeout=1s` ensures Nginx retries on almost every request from `loading.html`
```
A small comment about styling. In general whenever we refer to something code-like, we always prefer to use backticks (like in Markdown). Probably not applied everywhere, but for sure all new code respects this convention.